docs(api): Revisions to the Examples page

jwwojak opened the pull request:
# Overview

This PR makes Flex-related code and text changes to the [Examples page](file:///Users/joewojak/Documents/GHE/opentrons2/api/docs/build/html/v2/new_examples.html) of the API docs. Making changes to make it less OT-2 specific and incorporate both robots into the docs.

Also, each code sample is complete top to bottom. This is a departure from previous pages where samples focus on particular snippets. The reason for the formality is we want each example to be full and complete and ready to go for copy/paste operations.

# Test Plan

The usual review process.

I tested code samples in the app.

Sandbox: http://sandbox.docs.opentrons.com/new-examples-revisions/v2/new_examples.html

# Changelog

There are some significant changes to this page such as:

- Title change to "Protocol Examples" from just "Examples"
- Add new intro
- Add new section w/ an explanation about the code to the intended audience -- mainly _inexperienced_ python users.
- Add new section (similar to tutorial) that describes the instruments and labware used throughout.
- Broke up the first piece, "Transferring Liquids" into 2 subsections: one that covers basic/building block commands and the other for the more "advanced" `transfer()` method.
- Using `.. tabs::` directive to provide Flex and OT-2 code samples.

# Review requests

Please review and comment. I think these are significant and substantial improvements to the page.

# Risk assessment

Low regarding text. Mildly irritating to users if the code examples are incorrect.

jwwojak committed:
Cleanup section headers and add equipment

jwwojak committed:
Intro, header styles, list labware, tabs

For Basic Transfer section.

jwwojak committed:
Transfer examples for Flex and OT-2

jwwojak committed:
Minot intro changes

jwwojak committed:
More intro text changes, minor

jwwojak committed:
Title, intro, and 1st section text changes.

jwwojak committed:
Merge branch 'edge' into new-examples-revisions

jwwojak committed:
Loop section text and code revisions

jwwojak committed:
Air gaps text and code revisions

jwwojak committed:
Dilution revised text and code samples

jwwojak committed:
Delete duplicated code from Dilutent section

jwwojak committed:
Revise plate mapping text and code

jwwojak committed:
New section, reorg 1st example section

jwwojak committed:
Pre-lunch save tweaks and changes

jwwojak committed:
One more small text edit

jwwojak committed:
Change code quote marks and small text mods

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
+*****************
 
-All examples on this page use a ``'corning_96_wellplate_360ul_flat'`` (`an ANSI standard 96-well plate <https://labware.opentrons.com/corning_96_wellplate_360ul_flat>`_) in slot 1, and two ``'opentrons_96_tiprack_300ul'`` (`the Opentrons standard 300 µL tiprack <https://labware.opentrons.com/opentrons_96_tiprack_300ul>`_) in slots 2 and 3. They also require a P300 Single attached to the right mou ...
+This page provides simple, ready-made protocols for Flex and OT-2. Feel free to copy and modify these examples to create unique protocols that help automate your laboratory workflows. Also, experimenting with these protocols is another way to build upon the skills you've learned from working through the :ref:`tutorial`. Try adding different hardware, labware, and commands to a sample protocol and ...
```
```suggestion
This page provides simple, ready-made protocols for Flex and OT-2. Feel free to copy and modify these examples to create unique protocols that help automate your laboratory workflows. Also, experimenting with these protocols is another way to build upon the skills you've learned from working through the :ref:`tutorial`. Try adding different hardware, labware, and commands to a sample protocol and test its validity after importing it into the Opentrons App.
```
🆙

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
-**************
-Basic Transfer
-**************
+These sample protocols are designed for anyone using an Opentrons Flex or OT-2 liquid handling robot. 
```
nitpick, this breaks really badly when formatted (all one line except `bot.`). maybe strike "liquid handling"

jwwojak replied:
@ecormany I don't see this break. There are 2 separate lines here, is that what you're referring to? I think this is ok.

Intro sentence:
> These sample protocols are designed for anyone using an Opentrons Flex or OT-2 liquid handling robot.

hard return, start next paragraph

> For our users with little to no Python experience, . . .

ecormany replied:
this, in the sandbox out-
put.

<img width="696" alt="image" src="https://github.com/Opentrons/opentrons/assets/1958332/4e46e0cd-dbf0-4ad0-9f2d-b0c0601b71c2">

jwwojak replied:
Fixing.

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
+These sample protocols are designed for anyone using an Opentrons Flex or OT-2 liquid handling robot. 
 
-Moving 100 µL from one well to another:
+For our users with little to no Python experience, we’ve taken some liberties with the syntax and structure of the code to make it easier to understand. For example, we’ve formatted the samples with line breaks to show method arguments clearly and to avoid horizontal scrolling. Additionally, the methods use named arguments (or `named parameters <https://en.wikipedia.org/wiki/Named_parameter>`_) i ...
```
I thought this was super-strict formatting, but you're right — Black will try to fit as much on one line even if that mixes and matches one-line vs. multi-line style.

But I think we should use Black parenthesis style, placing the closing parenthesis on a separate line, so:

```
    pipette_1 = protocol.load_instrument(
        instrument_name="flex_1channel_1000",
        mount="left",
        tip_racks=[tiprack_1]
    )
``` 

instead of 
```
        ...
        tip_racks=[tiprack_1])

jwwojak replied:
Oh, please, can we not do this? --- the dangling closing parens on it's own line.

We don't do this elsewhere in the docs. It looks weird and takes an extra line. I really think we shouldn't do this, unless it's part of a top to bottom, front to back revision of the code examples for the entire docs site.

ecormany replied:
ooooh, we have a mix elsewhere. yeah, we don't have to change it now. for a rainy day ☔

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
+These sample protocols are designed for anyone using an Opentrons Flex or OT-2 liquid handling robot. 
 
-Moving 100 µL from one well to another:
+For our users with little to no Python experience, we’ve taken some liberties with the syntax and structure of the code to make it easier to understand. For example, we’ve formatted the samples with line breaks to show method arguments clearly and to avoid horizontal scrolling. Additionally, the methods use named arguments (or `named parameters <https://en.wikipedia.org/wiki/Named_parameter>`_) i ...
```
```suggestion
For our users with little to no Python experience, we’ve taken some liberties with the syntax and structure of the code to make it easier to understand. For example, we’ve formatted the samples with line breaks to show method arguments clearly and to avoid horizontal scrolling. Additionally, the methods use `named arguments <https://en.wikipedia.org/wiki/Named_parameter>`_ instead of positional arguments. For example::
```
i like the wiki link, and it includes "named arguments" in bold in the first line, so we don't have to equivocate here.

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
+    # This code uses named arguments
+    tiprack_1 = protocol.load_labware(
+        load_name='opentrons_flex_96_tiprack_200ul',
+        location='D2')
```
here's an example of where i'd move the parenthesis down. i'm not going to mark all of them.

jwwojak replied:
I think we ought not to do this. The code works, we don't do it elsewhere, and it looks weird.

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
+    tiprack_1 = protocol.load_labware('opentrons_flex_96_tiprack_200ul','D2')   
 
-    metadata = {'apiLevel': '|apiLevel|'}
+Both examples instantiate the variable ``tiprack_1`` with a Flex tip rack, but the former is more explicit. It shows the parameter name and it's value together (e.g. ``location='D2'``), which may be helpful when you're unsure about what's going on in a protocol code sample.
```
```suggestion
Both examples instantiate the variable ``tiprack_1`` with a Flex tip rack, but the former is more explicit. It shows the parameter name and its value together (e.g. ``location='D2'``), which may be helpful when you're unsure about what's going on in a protocol code sample.
```
😇

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
+The sample protocols all use the following pipettes:
 
-This accomplishes the same thing as the following basic commands:
+* Flex 1-Channel Pipette (5-1000 µL). The API load name for this pipette is ``flex_1channel_1000``. 
```
```suggestion
* Flex 1-Channel Pipette (5–1000 µL). The API load name for this pipette is ``flex_1channel_1000``. 
```
i like en dashes

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
 
-This accomplishes the same thing as the following basic commands:
+* Flex 1-Channel Pipette (5-1000 µL). The API load name for this pipette is ``flex_1channel_1000``. 
+* P300 Single-Channel GEN2 pipette (20-300 µL) for the OT-2. The API load name for this pipette is ``p300_single_gen2``. 
```
```suggestion
* P300 Single-Channel GEN2 pipette for the OT-2. The API load name for this pipette is ``p300_single_gen2``. 
```
the range isn't part of the product name here (it is for Flex pipettes)

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
+                plate = protocol.load_labware(
+                load_name='corning_96_wellplate_360ul_flat',
+                location=1)
+            tiprack_1 = protocol.load_labware(
+                load_name='opentrons_96_tiprack_300ul',
+                location=2)
+            p300 = protocol.load_instrument(
+                instrument_name='p300_single',
+                mount='left',
+                tip_racks=[tiprack_1])
+
+            p300.pick_up_tip()
+            p300.aspirate(100, plate['A1'])
+            p300.dispense(100, plate['B1'])
+            p300.drop_tip()
```
\+ 1 tab indent

jwwojak replied:
Yikes. Fixing.

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
+Advanced Method
+---------------
+
+This protocol accomplishes the same thing as the previous example, but does it a little more efficiently. Notice how it uses the :py:meth:`.InstrumentContext.transfer` method to move liquid between well plates. The ``aspirate`` and ``dispense`` arguments are part of ``transfer()`` method parameters. 
```
The arguments are `source` and `dest`, not aspirate and dispense.
Or, both aspirating and dispensing are performed as part of the transfer command.

jwwojak replied:
I think I mean that the actions of aspirating and dispensing were taken care of by `transfer` and you didn't need to call them.  My mistake. Fixing.

> The source and destination well arguments (e.g., `plate['A1'], plate['B1']`) are part of `transfer()` method parameters. You don’t need separate calls to `aspirate` or `dispense` here.

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
+                    tip_racks=[tiprack_1])
+                # transfer 100 µL from well A1 to well B1
+                pipette_1.transfer(100, plate['A1'], plate['B1'])
+                pipette_1.drop_tip()
```
```suggestion
```
Delete this line or protocol won't analyze. `transfer()` handles tip pickup and drop.

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
+                mount='left',
+                tip_racks=[tiprack_1])
+            # transfer 100 µL from well A1 to well B1
+            p300.transfer(volume=100, plate['A1'], plate['B1'])
+            p300.drop_tip()
```
```suggestion
            p300.transfer(100, plate['A1'], plate['B1'])
```
Python error if positional argument follows named argument. Also removes drop tip for same reason as in Flex example.

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
+                for i in range(8):
+                    p300.distribute(200, reservoir.wells()[i], plate.rows()[i])
+
+Notice here how `Python's range class <https://docs.python.org/3/library/stdtypes.html#range>`_ (e.g., ``range(8)``) determines how many times the code loops. Also, in Python, a range of numbers is *exclusive* of the end value and counting starts at 0, not 1. For the Corning 96-well plate used here, this means well A1=0, B1=1, C1=2, and so on to the last well in the row, which is H1=7.
```
```suggestion
Notice here how Python's :py:class:`range` class (e.g., ``range(8)``) determines how many times the code loops. Also, in Python, a range of numbers is *exclusive* of the end value and counting starts at 0, not 1. For the Corning 96-well plate used here, this means well A1=0, B1=1, C1=2, and so on to the last well in the row, which is H1=7.
```
Sphinx can directly reference the Python standard library.

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
-            # save the source well and destination column to variables
-            source = reservoir.wells()[i]
-            row = plate.rows()[i]
+Notice here how Python's `slice functionality <https://docs.python.org/3/library/functions.html#slice>`_ (in the code sample as ``[:4]``) lets us select the first five wells of the well plate only. Also, in Python, a range of numbers is *exclusive* of the end value and counting starts at 0, not 1. For the Corning 96-well plate used here, this means well A1=0, B1=1, C1=2, and so on to the last wel ...
```
```suggestion
Notice here how Python's :py:class:`slice` functionality (in the code sample as ``[:4]``) lets us select the first five wells of the well plate only. Also, in Python, a range of numbers is *exclusive* of the end value and counting starts at 0, not 1. For the Corning 96-well plate used here, this means well A1=0, B1=1, C1=2, and so on to the last well used, which is E1=4. See also, the :ref:`tutorial-commands` section of the Tutorial.
```
sphinx-ify here too

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
+Dilution
+========
+
+This protocol dispenses diluent to all wells of a Corning 96-well plate. Next, it dilutes eight samples from the reservoir across all 8 columns of the plate.
```
```suggestion
This protocol dispenses diluent to all wells of a Corning 96-well plate. Next, it dilutes 8 samples from the reservoir across all 8 columns of the plate.
```
stick with numerals throughout since we have big and small numbers in one sentence

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
+Notice here how Python's `slice functionality <https://docs.python.org/3/library/functions.html#slice>`_ (in the code sample as ``[:4]``) lets us select the first five wells of the well plate only. Also, in Python, a range of numbers is *exclusive* of the end value and counting starts at 0, not 1. For the Corning 96-well plate used here, this means well A1=0, B1=1, C1=2, and so on to the last wel ...
 
-            # transfer 30uL of source to first well in column
+Dilution
```
we _could_ remove this section entirely because it's very much the basis for the tutorial. your call.

jwwojak replied:
I'd err on the side of keeping it.

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
+                tiprack_2 = protocol.load_labware(
+                    load_name='opentrons_96_tiprack_300ul',
+                    location=3)
+            reservoir = protocol.load_labware(
+                    load_name='usascientific_12_reservoir_22ml',
+                    location=4)
+            p300 = protocol.load_instrument(
+                    instrument_name='p300_single',
+                    mount='right',
+                    tip_racks=[tiprack_1, tiprack_2])
+            # Dispense diluent
+            p300.distribute(50, reservoir['A12'], plate.wells())
+
+            # loop through each row
+            for i in range(8):
+                # save the source well and destination column to variables
+                source = reservoir.wells()[i]
+                source = reservoir.wells()[i]
+                row = plate.rows()[i]
+
+            # transfer 30 µL of source to first well in column
             p300.transfer(30, source, row[0], mix_after=(3, 25))
 
             # dilute the sample down the column
             p300.transfer(
                 30, row[:11], row[1:],
                 mix_after=(3, 25))
```
+1 tab indent here

jwwojak replied:
oops. Fixing that now.

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
+                    89, 90, 91, 92, 93, 94, 95, 96
+                    ]
+
+                    pipette_1.distribute(water_volumes, reservoir['A12'], plate.wells())
```
```suggestion
                pipette_1.distribute(water_volumes, reservoir['A12'], plate.wells())
```
-1 tab indent

ecormany commented on api/docs/v2/new_examples.rst:
```diff
@@ -2,187 +2,520 @@
+                    89, 90, 91, 92, 93, 94, 95, 96
+                    ]
+
+                    p300.distribute(water_volumes, reservoir['A12'], plate.wells())
```
```suggestion
                p300.distribute(water_volumes, reservoir['A12'], plate.wells())
```
outdent here too

ecormany reviewed (changes requested):
Overall structure improvements are good! Changes are mostly to get the protocol code analyzing.

jwwojak committed:
Apply suggestions from code review

Committing reviewer recommendations.

Co-authored-by: Ed Cormany <edward.cormany@opentrons.com>

jwwojak committed:
Fixing code indentation

Lines 405-421

jwwojak committed:
Code sample indentation fix

jwwojak committed:
Remove line break in Using Protocols section

Though this was deliberate. Will remove to make one paragraph.

jwwojak committed:
Advanced Method - fix code description

I mistakenly wrote that the .transfer method arguments were 'aspirate' and 'dispense'. Should be 'source' and 'dest'

jwwojak commented:
Accepted all batch revisions and made changes based on open review comments. Re-requesting review. Woo.

ecormany committed:
final indentation fixes

ecormany reviewed (approved):
Added a little more meaningful whitespace. Now everything is good to go!
